src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,213 @@
+
+        writer.Init(std::move(handle), false);
+
+        ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag, *apData));
+        ReturnErrorOnFailure(writer.Finalize(&handle));
+
+        //
+        // Compact the buffer down to a more reasonably sized packet buffer
+        // if we can.
+        //
+        handle.RightSize();
+
+        mBufferedList.push_back(std::move(handle));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BufferedReadCallback::DispatchBufferedData(const ReadClient * apReadClient, const ConcreteAttributePath & aPath,
+                                                      const StatusIB & aStatusIB, bool endOfReport)
+{
+    if (aPath == mBufferedPath)
+    {
+        //
+        // If we encountered the same list again and it's not the last DataIB, then
+        // we need to continue to buffer up this list's data, so return immediately without dispatching
+        // the existing buffered up contents.
+        //
+        if (!endOfReport)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        //
+        // If we had previously buffered up data for this list and now we have encountered
+        // an error for this list, that error takes precedence and the buffered data is now
+        // rendered invalid. Return immediately without dispatching the existing buffered up contents.
+        //
+        if (aStatusIB.mStatus != Protocols::InteractionModel::Status::Success)
+        {
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    if (!mBufferedPath.IsListOperation())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    StatusIB statusIB;
+    System::PacketBufferTLVReader reader;
+
+    ReturnErrorOnFailure(GenerateListTLV(reader));
+
+    //
+    // Update the list operation to now reflect the delivery of the entire list
+    // i.e a replace all operation.
+    //
+    mBufferedPath.mListOp = ConcreteAttributePath::ListOperation::ReplaceAll;
+
+    //
+    // Advance the reader forward to the list itself
+    //
+    reader.Next();
+
+    mCallback.OnAttributeData(apReadClient, mBufferedPath, &reader, statusIB);
+
+    //
+    // Clear out our buffered contents to free up allocated buffers
+    //
+    mBufferedList.clear();
+
+    return CHIP_NO_ERROR;
+}
+
+void BufferedReadCallback::OnAttributeData(const ReadClient * apReadClient, const ConcreteAttributePath & aPath,
+                                           TLV::TLVReader * apData, const StatusIB & aStatus)
+{
+    //
+    // First, let's dispatch to our registered callback any buffered up list data from previous calls.
+    //
+    DispatchBufferedData(apReadClient, aPath, aStatus);
+
+    //
+    // We buffer up list data (only if the status was successful)
+    //
+    if (aPath.IsListOperation() && aStatus.mStatus == Protocols::InteractionModel::Status::Success)
+    {
+        ReturnOnFailure(BufferData(aPath, apData));
```

Comment:
should we log failure as well or make this method return a CHIP_ERROR? silent error seems dangerous.